File: validation/surface_cases.jl
using Arrhenius
using NPZ
using Test
using TOML

include(joinpath(@__DIR__, "..", "example", "reactors", "surface_solver.jl"))

function validate_surfaces(directory)
    reports = Dict{String,Any}()
    for name in ("pt_h2", "diamond")
        m = SurfaceMechanism(joinpath(directory, "$name.surface.npz"))
        ref = npzread(joinpath(directory, "$name.reference.npz"))
        errors = Dict{String,Float64}()
        for (key, getter) in (("forward", w->w.forward_rates), ("reverse", w->w.reverse_rates),
                              ("forward_constants", w->w.forward_rate_constants),
                              ("reverse_constants", w->w.reverse_rate_constants),
                              ("production", w->w.production_rates))
            predicted = similar(ref[key])
            for j in eachindex(ref["temperatures"])
                s = IdealSurface(m; temperature=ref["temperatures"][j],
                    pressure=only(ref["pressure"]), gas_temperature=ref["gas_temperatures"][j],
                    mole_fractions=ref["mole_fractions"], coverages=ref["coverages"][:,j])
                predicted[:,j] = getter(surface_rates!(SurfaceWorkspace(m), s))
            end
            scale = max.(abs.(ref[key]), 1e-30)
            errors[key*"_relative_error"] = maximum(abs.(predicted-ref[key])./scale)
            @test isapprox(predicted, ref[key]; rtol=2e-11, atol=1e-25)
        end
        s = IdealSurface(m; temperature=only(ref["temperature"]), pressure=only(ref["pressure"]),
                         mole_fractions=ref["mole_fractions"], coverages=ref["initial_coverages"])
        times = vec(ref["times"])
        solution = solve_surface(s, (0.0,last(times)); integrator=native_surface_bdf,
            reltol=1e-10, abstol=1e-18, saveat=times, tstops=times[2:end], maxiters=1_000_000)
        trajectory = reduce(hcat, solution.u)
        errors["trajectory_absolute_error"] = maximum(abs, trajectory-ref["trajectory"])
        errors["site_balance_drift"] = maximum(abs, vec(sum(trajectory; dims=1)).-1)
        @test errors["trajectory_absolute_error"] < 2e-7
        @test errors["site_balance_drift"] < 1e-9
        @test minimum(trajectory) >= -1e-13
        stationary = steady_coverages(s; integrator=native_surface_bdf,
            reltol=1e-11, abstol=1e-19, maxiters=1_000_000, initial_coverages=solution.u[end])
        errors["steady_coverage_error"] = maximum(abs, stationary.coverages-ref["steady_coverages"])
        errors["steady_residual_per_second"] = stationary.residual
        @test errors["steady_coverage_error"] < 2e-7
        @test abs(sum(stationary.coverages)-1) < 1e-9
        rates = surface_rates(s, stationary.coverages)
        @test maximum(abs, rates.elemental_rates) < 1e-13
        @test isapprox(vcat(rates.surface,rates.gas,rates.bulk), ref["steady_production"]; rtol=1e-5, atol=1e-13)
        reports[name] = errors
        @show name errors
        npzwrite(joinpath(directory, "$name.native.npz"), Dict("trajectory"=>trajectory,
            "steady_coverages"=>stationary.coverages, "times"=>times))
    end
    open(joinpath(directory, "surface_validation.toml"), "w") do io
        TOML.print(io, reports; sorted=true)
    end
    return reports
end

if abspath(PROGRAM_FILE) == @__FILE__
    length(ARGS) == 1 || error("usage: julia surface_cases.jl REFERENCE_DIRECTORY")
    @testset "Cantera surface kinetics and coverage references" begin
        validate_surfaces(ARGS[1])
    end
end

File: validation/surface_cases.py
"""Cantera surface-rate, coverage-trajectory, and stationary-state references.

The Pt/H2 point is the catalytic_combustion.py surface initialization at 900 K
and one atmosphere. The diamond point follows diamond_cvd.py at 1200 K, 20 Torr.
Additional positive coverage vectors and unequal gas/surface temperatures are
selected regression points, not published operating ranges.
https://cantera.org/dev/examples/python/onedim/catalytic_combustion.html
https://cantera.org/dev/examples/python/kinetics/diamond_cvd.html
"""
from __future__ import annotations

import argparse
import importlib.util
import json
from pathlib import Path

import cantera as ct
import numpy as np


def main(output):
    output.mkdir(parents=True, exist_ok=True)
    spec = importlib.util.spec_from_file_location("export_surface", Path(__file__).resolve().parents[1]/"mechanism"/"export_surface.py")
    exporter = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(exporter)
    cases = [
        ("pt_h2", "ptcombust.yaml", "Pt_surf", 900.0, ct.one_atm,
         "H2:0.05,O2:0.21,N2:0.78,AR:0.01", {"PT(S)":0.5,"O(S)":0.5}),
        ("diamond", "diamond.yaml", "diamond_100", 1200.0, 20*ct.one_atm/760,
         "H:0.002,H2:0.988,CH3:0.0002,CH4:0.01", None),
    ]
    manifest = {"cantera_version":ct.__version__, "cases":{}}
    for name, mechanism, phase, T, P, composition, theta0 in cases:
        surface, gas_file = exporter.export_surface(mechanism, phase, output/f"{name}.surface.npz")
        gas = next(p for p in surface.adjacent.values() if p.thermo_model == "ideal-gas")
        surface.TP = T, P
        gas.TPX = T, P, composition
        if theta0 is not None:
            surface.coverages = theta0
        theta0 = surface.coverages.copy()
        names = surface.species_names + gas.species_names + [sp for p in surface.adjacent.values() if p is not gas for sp in p.species_names]
        order = [surface.kinetics_species_index(sp) for sp in names]
        rng = np.random.default_rng(4127)
        theta = [theta0]
        for k in range(8):
            q = np.exp(rng.uniform(-12, 1, surface.n_species))
            theta.append(q/q.sum())
        temperatures = np.array([T, 450, 700, 900, 1000, 1100, 1400, T, T],float)
        gas_temperatures = temperatures.copy()
        gas_temperatures[-2:] = [300, 1300]
        ropf, ropr, kf, kr, wdot = [], [], [], [], []
        for q, Ts, Tg in zip(theta, temperatures, gas_temperatures):
            surface.TP = Ts, P
            gas.TPX = Tg, P, composition
            surface.coverages = q
            for p in surface.adjacent.values():
                if p is not gas:
                    p.TP = Ts, P
            ropf.append(surface.forward_rates_of_progress.copy())
            ropr.append(surface.reverse_rates_of_progress.copy())
            kf.append(surface.forward_rate_constants.copy())
            kr.append(surface.reverse_rate_constants.copy())
            wdot.append(surface.net_production_rates[order].copy())
        surface.TP = T, P
        gas.TPX = T, P, composition
        for p in surface.adjacent.values():
            if p is not gas:
                p.TP = T, P
        surface.coverages = theta0
        times = np.concatenate(([0.0], np.geomspace(1e-12, 1.0, 121)))
        trajectory = [surface.coverages.copy()]
        for left, right in zip(times[:-1], times[1:]):
            surface.advance_coverages(right-left, rtol=1e-11, atol=1e-20, max_steps=100000)
            trajectory.append(surface.coverages.copy())
        surface.advance_coverages_to_steady_state()
        steady = surface.coverages.copy()
        production_steady = surface.net_production_rates[order].copy()
        np.savez(output/f"{name}.reference.npz", temperatures=temperatures,
            gas_temperatures=gas_temperatures, pressure=np.array([P]),
            temperature=np.array([T]), mole_fractions=gas.X.copy(),
            coverages=np.array(theta).T, forward=np.array(ropf).T,
            reverse=np.array(ropr).T, forward_constants=np.array(kf).T,
            reverse_constants=np.array(kr).T, production=np.array(wdot).T,
            times=times, trajectory=np.array(trajectory).T, initial_coverages=theta0,
            steady_coverages=steady, steady_production=production_steady)
        manifest["cases"][name] = {"mechanism":mechanism,"phase":phase,"temperature":T,
            "pressure":P,"composition":composition,"rate_states":len(theta),
            "trajectory_points":len(times),"gas_file":gas_file.name}
    (output/"manifest.json").write_text(json.dumps(manifest,indent=2)+"\n")


if __name__ == "__main__":
    parser=argparse.ArgumentParser(description=__doc__)
    parser.add_argument("output",type=Path)
    main(parser.parse_args().output)
